sources/imdsclient/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+/*!
+The imdsclient library provides high-level methods to interact with the AWS Instance Metadata Service.
+The high-level methods provided are [`fetch_dynamic`], [`fetch_metadata`], and [`fetch_userdata`].
+
+For more control, and to query IMDS without high-level wrappers, there is also a [`fetch_imds`] method.
+This method is useful for specifying things like a pinned date for the IMDS schema version.
+*/
+
+#![deny(rust_2018_idioms)]
+
+use http::StatusCode;
+use log::{debug, info, trace, warn};
+use reqwest::blocking::Client;
+use snafu::{ensure, ResultExt};
+
+const IMDS_BASE_URI: &str = "http://169.254.169.254";
+const IMDS_PINNED_DATE: &str = "2020-10-27";
+
+// Currently only able to get fetch session tokens from `latest`
+const IMDS_SESSION_TOKEN_URI: &str = "http://169.254.169.254/latest/api/token";
+
+/// A client for making IMDSv2 queries.
+/// It obtains a session token when it is first instantiated and is reused between helper functions.
+pub struct ImdsClient {
+    client: Client,
+    session_token: String,
+}
+
+impl ImdsClient {
+    pub fn new() -> Result<Self> {
+        let client = Client::new();
+        let session_token = fetch_token(&client)?;
+        Ok(Self {
+            client,
+            session_token,
+        })
+    }
+
+    /// Helper to fetch `dynamic` targets from IMDS, preferring an override file if present.
+    pub fn fetch_dynamic(&mut self, end_target: &str, description: &str) -> Result<Option<String>> {
```

Comment:
Is fetch_dynamic a local helper or something we expect client programs to use?

Reply:
The only case I know of is that `early-boot-config` needs it to [fetch the region](https://github.com/bottlerocket-os/bottlerocket/blob/bf2a78ab8c8e628d8b2c6365bdce3958b0635d4f/sources/api/early-boot-config/src/provider/aws.rs#L149) for the API.

Reply:
OK, if it's being used then pub is fine. I was just looking for pub lock-down opportunities.